### PR TITLE
Implement `dns` component and Components Refactor

### PR DESCRIPTION
**Pull request by mateoflorido**
## Overview
This Pull Request introduces a mechanism for separating logic for component management, allowing for the configuration of these components. Additionally, the DNS component is implemented.
## Changes
- Introduced custom handlers for the `dns` and `network` components.
- Refactored the disable logic.
- Refactored the `network` component to adhere to the new management logic.
- Updated CoreDNS to version 1.29.0.

## Manual Testing
- Pods
```
NAMESPACE     NAME                               READY   STATUS    RESTARTS   AGE
kube-system   cilium-operator-56bdb99ff6-j5w2q   1/1     Running   0          8m5s
kube-system   cilium-v4q8z                       1/1     Running   0          8m5s
kube-system   ck-dns-coredns-7f648ff498-z7zrj    1/1     Running   0          8m
```
- nslookup
```
root@k8s-test:~# /snap/k8s/current/bin/kubectl --kubeconfig /var/snap/k8s/common/etc/kubernetes/admin.conf run --rm -it --image alpine -- sh
If you don't see a command prompt, try pressing enter.
/ # nslookup canonical.com
Server:		10.152.183.10
Address:	10.152.183.10:53

Non-authoritative answer:
Name:	canonical.com
Address: 2620:2d:4000:1::28
Name:	canonical.com
Address: 2620:2d:4000:1::26
Name:	canonical.com
Address: 2620:2d:4000:1::27

Non-authoritative answer:
Name:	canonical.com
Address: 185.125.190.21
Name:	canonical.com
Address: 185.125.190.29
Name:	canonical.com
Address: 185.125.190.20

/ # nslookup google.com
Server:		10.152.183.10
Address:	10.152.183.10:53

Non-authoritative answer:
Name:	google.com
Address: 2800:3f0:4005:411::200e

Non-authoritative answer:
Name:	google.com
Address: 142.250.218.142
```
## Notes
- Testing will be introduced in a follow-up PR.
- Custom configuration will be introduced once the machinery to provide custom configuration is in place.

### CODE REVIEW

File: src/k8s/pkg/component/hook.go
Comment (neoaggelos) [inline src/k8s/pkg/component/hook.go:27]: this should be handled with the kubelet patch, so no need to adjust it

File: src/k8s/pkg/component/hook.go
Comment (neoaggelos) [inline src/k8s/pkg/component/hook.go:44]: Since we will be retrieving the service IP later on from the apiserver, we can avoid this entirely.

File: src/k8s/pkg/component/postconfig.go
Comment (neoaggelos) [inline src/k8s/pkg/component/postconfig.go:32]: i wonder if we should be simply automatically setting `--cluster-domain` in the default args of kubelet. that way, we would only have to set `--cluster-dns` in this place.

File: src/k8s/pkg/utils/file.go
Comment (neoaggelos) [inline src/k8s/pkg/utils/file.go:101]: please see https://github.com/canonical/k8s-snap/pull/49/files#diff-506efc1ed4c8530427a69550ef2bbc3caba657b9a33dde1851cc3b1f07ecf17aR39 (there was also some discussion around why want to avoid a regex for updating arguments in the same PR).

My preference would be to wait until #49 is merged and adjust the code. We need to move #49 forward.

**Review by neoaggelos — COMMENTED** [review]
Left some initial comments. Good work overall, but I believe we will have to do some adjustments on this PR before it goes in.

File: src/k8s/pkg/utils/file.go
Comment (mateoflorido) [reply src/k8s/pkg/utils/file.go:101]: I see, I'll wait until #49 is merged to refactor the PR. Thanks for the heads up :)

**Comment by mateoflorido** [comment]
Using custom nameservers and service ip:
```
root@k8s-test:~# k8s enable dns --upstream-nameservers=1.1.1.1,8.8.8.8 --service-ip=10.152.183.11
DNS component enabled
root@k8s-test:~# /snap/k8s/current/bin/kubectl --kubeconfig /var/snap/k8s/common/etc/kubernetes/admin.conf run --rm -it --image alpine -- sh
If you don't see a command prompt, try pressing enter.
/ # nslookup canonical.com
Server:		10.152.183.11
Address:	10.152.183.11:53

Non-authoritative answer:
Name:	canonical.com
Address: 185.125.190.21
Name:	canonical.com
Address: 185.125.190.29
Name:	canonical.com
Address: 185.125.190.20

Non-authoritative answer:
Name:	canonical.com
Address: 2620:2d:4000:1::27
Name:	canonical.com
Address: 2620:2d:4000:1::26
Name:	canonical.com
Address: 2620:2d:4000:1::28
```

File: src/k8s/api/v1/component.go
Comment (neoaggelos) [inline src/k8s/api/v1/component.go:19]: you can use the `ComponentStatus` alias here, like in `UpdateComponentRequest`.

File: src/k8s/cmd/k8s/k8s_enable_dns.go
Comment (neoaggelos) [inline src/k8s/cmd/k8s/k8s_enable_dns.go:37]: Use the "enum"-like `api.ComponentEnable` here

File: src/k8s/pkg/component/component.go
Comment (neoaggelos) [inline src/k8s/pkg/component/component.go:20]: We fully control callers here, so let's just change `Enable`?

```suggestion
	// Enable enables a k8s component, optionally specifying custom configuration options.
	Enable(name string, values map[string]any) error
```

File: src/k8s/pkg/component/dns.go
Comment (neoaggelos) [inline src/k8s/pkg/component/dns.go:13]: A minor note on this. `EnableDNSComponent` is internal machinery, whereas `api.UpdateDNSComponentRequest` is "public" API.

If we had to change the API in any way (or this implementation), we don't want to have to make this change in multiple places. As such, I would recommend changing the definition to:

```
func EnableDNSComponent(upstreamNameserver, clusterDomain, serviceIP string) error
```

This way, we can change the API (or the EnableDNSComponent) without having to track down changes across multiple files.

File: src/k8s/pkg/k8s/client/component.go
Comment (neoaggelos) [inline src/k8s/pkg/k8s/client/component.go:35]: Side-note, when wrapping errors we should add context that the function that failed did not have, e.g.

```go
return fmt.Errorf("enable dns component failed: %w", err)
```

File: src/k8s/pkg/k8sd/api/component.go
Comment (neoaggelos) [inline src/k8s/pkg/k8sd/api/component.go:26]: thanks for keeping the todo here and in the client code :+1:

File: src/k8s/pkg/utils/dns.go
Comment (neoaggelos) [inline src/k8s/pkg/utils/dns.go:40]: Do we still need this here? Kubelet should be able to handle the default resolv.conf already.

**Review by neoaggelos — COMMENTED** [review]
Excellent work @mateoflorido! Leaving some follow-up comments

File: src/k8s/pkg/utils/kubectl.go
Comment (bschimke95) [inline src/k8s/pkg/utils/kubectl.go:1]: seems like we have done a bit of duplicate work here:
https://github.com/canonical/k8s-snap/pull/57/files

I would not reimplement the go-k8s library but rather - if possible - provide more abstraction.
For example, from my PR: Not implementing "k8s.GetNodes" and then iterate on the node objects but rather "k8s.ClusterReady".

**Review by bschimke95 — COMMENTED** [review]
Great work @mateoflorido 
Left some comments.

File: src/k8s/cmd/k8s/k8s_disable.go
Comment (neoaggelos) [inline src/k8s/cmd/k8s/k8s_disable.go:20]: This looks great, but I would recommend having separate commands here as well. It is much easier to be able to specify custom arguments in the future (if we have to), same as we have in the enable command.

File: src/k8s/cmd/k8s/k8s_enable_dns.go
Comment (neoaggelos) [inline src/k8s/cmd/k8s/k8s_enable_dns.go:11]: Just a nit, same as other commands, `enableDNSCmdConfig`

File: src/k8s/pkg/component/component.go
Comment (neoaggelos) [inline src/k8s/pkg/component/component.go:19]: if `Enable(name)` is equivalent to `EnableWithValues(name, nil)`, then let's only have a single one and update all callers:

```go
Enable(name string, values map[string]any) error
```

**Review by neoaggelos — COMMENTED** [review]
Almost there, nice work @mateoflorido! Leaving a few more comments

File: src/k8s/pkg/component/dns.go
Comment (neoaggelos) [inline src/k8s/pkg/component/dns.go:13]: nit; as a matter of precedence, make `s snap.Snap` the first argument.

File: src/k8s/pkg/component/component.go
Comment (neoaggelos) [inline src/k8s/pkg/component/component.go:87]: yes, see `helmClient` does not implement the `ComponentManager` with this. A common approach to ensure we don't break the interface is to add the following at the end of the file, which ensures that a `*helmClient` can be assigned to a `ComponentManager` interface:

```go
var _ ComponentManager = &helmClient{}
```

File: src/k8s/pkg/utils/kubectl.go
Comment (bschimke95) [inline src/k8s/pkg/utils/kubectl.go:1]: With that, we have two utils doing the same thing:
https://github.com/canonical/k8s-snap/tree/main/src/k8s/pkg/utils/k8s

Please merge them into one.

**Review by bschimke95 — COMMENTED** [review]
Great work Mateo, two final comments

File: src/k8s/pkg/component/dns.go
Comment (neoaggelos) [inline src/k8s/pkg/component/dns.go:28]: Can we also set `service.name=coredns` and `deployment.name=coredns`? Will reduce some of the stutter in the deployed names.

- https://github.com/coredns/helm/blob/master/charts/coredns/values.yaml#L60
- https://github.com/coredns/helm/blob/master/charts/coredns/values.yaml#L376

File: src/k8s/pkg/component/dns.go
Comment (neoaggelos) [inline src/k8s/pkg/component/dns.go:78]: These are part of the cluster config. Can you add a TODO to store them with `database.SetClusterConfig`?

I'm okay with not delaying this PR any longer, we will return to it at a later time

File: tests/e2e/tests/test_dns.py
Comment (neoaggelos) [inline tests/e2e/tests/test_dns.py:76]:
```suggestion
    assert "can't resolve" not in result.stdout.decode()
```

**Review by neoaggelos — COMMENTED** [review]
Just a couple final nits, great work!

**Review by bschimke95 — APPROVED** [review]
LGTM (when CI passes)

**Pull request merged**
